New Performance Trials and Optimizations

josharnold52 opened the pull request:
This request contains several new performance trials and some very significant optimizations.   I'll discuss the results of the new trials in #60.  

Note that my earlier `Zipper` refactor hurt selection times compared to the very high bar @ncreep had set.  These optimizations bring the performance back to that level and beyond.

Performance Tests:
- Added several new trials to measure various Zipper operations as well as overall "select/modify/unselect" times
- Added new implementations to selection trials to isolate the effect of the Zipper and of the bloom filter on selection times
- Moved Performance into its own package and split it into multiple classes
- New command line options to run certain "categories" of trials (load, selection, modify, etc.) 

Optimizations:
- Put hole information in its own tree structure (`ZipperHoleMap`) rather than a SortedMap
- Defer expensive zipper calculations to `unselect`
- Avoid Int boxing in non-transient path structures (`ZipperPath`)
- Use VectorCase wherever small vectors are likely
- Make PathCreator.PathVals traverse selections lazily
- Reuse empty Group instances (helps with size and with the bloom filter calculation)

This pull request does contain a large number files, but most of them are in the `test` directory.   Many of them are the result of splitting out the Performance class.

josharnold52 committed:
Split out Performance class into its own package.  Added new performance trials

josharnold52 committed:
Zipper Optimizations

josharnold52 committed:
Make PathCreator produce lazy Traversables rather than realized Seqs

josharnold52 committed:
Reuse the same empty Group instance.  This is an easy win for both memory and bloom filter creation costs

josharnold52 commented:
Some selected performance measurements.   The "select" trials are the same ones we've been using all along.  The "modify" trials measure an entire "select/modify/unselect" cycle.   They use the same selection as the "select" trials and add an attribute to the selected elements.

```
-- Execution Time --
[shallowSelectSmall]        Shallow selection in a 7 MB tree                   [378832 nodes, 133010 elems, 2772 attrs]
 + anti-xml:                    min:   3 ms, max:   9 ms, average:   4 ms          [Selected 8 nodes]
 + scala.xml:                   min:  47 ms, max:  49 ms, average:  47 ms          [Selected 8 nodes]
 + javax.xml:                   min:  17 ms, max:  18 ms, average:  17 ms          [Selected 8 nodes]

[deepSelectSmall]           Deep selection in a 7 MB tree                      [378832 nodes, 133010 elems, 2772 attrs]
 + anti-xml:                    min:   2 ms, max:   2 ms, average:   2 ms          [Selected 8 nodes]
 + scala.xml:                   min: 262 ms, max: 281 ms, average: 268 ms          [Selected 8 nodes]
 + javax.xml:                   min:  11 ms, max:  13 ms, average:  11 ms          [Selected 8 nodes]

[shallowModifyFewSmall]     shallow modify a few elements of a 7 MB tree       [modify 8 nodes]
 + anti-xml:                    min:   7 ms, max:   8 ms, average:   7 ms          [result has 378832 nodes, 133010 elems, 2780 attrs]
 + scala.xml:                   min:  35 ms, max:  36 ms, average:  35 ms          [result has 378832 nodes, 133010 elems, 2780 attrs]
 + javax.xml:                   min:  57 ms, max:  60 ms, average:  58 ms          [result has 378832 nodes, 133010 elems, 2780 attrs]

[deepModifyFewSmall]        deep modify a few elements of a 7 MB tree          [modify 8 nodes]
 + anti-xml:                    min:   2 ms, max:   3 ms, average:   2 ms          [result has 378832 nodes, 133010 elems, 2780 attrs]
 + scala.xml:                   min:  57 ms, max:  61 ms, average:  58 ms          [result has 378832 nodes, 133010 elems, 2780 attrs]
 + javax.xml:                   min:  65 ms, max:  71 ms, average:  67 ms          [result has 378832 nodes, 133010 elems, 2780 attrs]
```

josharnold52 committed:
Zipper.unselect should only increase the update times of nodes receiving updates.

josharnold52 committed:
Add explicit test for how unselect sets the update times of its result

josharnold52 committed:
Fix flatMap's setting of update times.

josharnold52 commented:
Ugh.  I should learn never to send a pull request after 3AM.   The above 3 commits are minor fixes to how update times are affected by various zipper operations.   I've added tests as well.

josharnold52 commented:
I've found some easy optimizations having to do with the bloom filter.  Shall I tack them on to this request?  The optimizations are:
1.  Groups that do not have Elems can reuse the same "empty" BloomFilter.  This improves the "cold" select trials by about 30% and should decrease the memory footprint of the filter.
2.  The element-by-name selector can pre-compute its bloom hash and reuse it.   This almost completely eliminates the bloom filter's overhead in the example I discussed in #60.  It also takes the `deepSelectSmall` trial down to 1ms on my device :-)

ncreep commented:
Very impressive!

A few notes/questions:
- When creating a broken `Zipper` you use `null`s for the irrelevant fields, wouldn't it be more informative to use `sys.error`?
- Is there a good reason to hold separate implementations for `flatMap` and `conditionalFlatMapWithIndex`'s instead of just lifting functions from `flatMap` into `conditionalFlatMapWithIndex` format?
- As the whole time thing is becoming more and more complicated, I think it would be a good a idea to define an "official time policy" for the `Zipper` to be used as guidelines for future updates.
- In the comment for the `additionalHoles` field you note that it may be used for data associated with actual nodes. As far as I can see, it is always implicitly assumed that the additional data is associated with an empty vector.    In any case, I think it may be useful in the future to have additional nodes hidden inside the `Zipper`.

josharnold52 commented:
> > When creating a broken Zipper you use nulls for the irrelevant fields, wouldn't it be more informative to use sys.error?

Perhaps so, but that would require using "def" and both of our earlier `updated` implementations had a bug having to do with chained defs.  That's an easy mistake to make, so I made all of the abstract Zipper fields "val"s to ensure it won't happen again.

However, I agree nulls are yucky.  I think the cleanest solution is to put the parent and all of the other fields in a separate "Context" class, and then make Zipper have a single val of type `Option[Context]`.  I'll go ahead and make that change.

Eventually I'd like to get rid of broken zippers altogether.   I've been doing some work along those lines, but I don't have anything ready for prime-time yet.

> > Is there a good reason to hold separate implementations for flatMap and conditionalFlatMapWithIndex's instead of just lifting functions from flatMap into conditionalFlatMapWithIndex format?

`conditionalFlatMapWithIndex` is optimized specifically for building Groups and Zippers whereas `flatMap` has to work with arbitrary builders.  

I forgot to mention `conditionalFlatMapWithIndex` when I described the changes.  It's probably the ugliest of the optimizations, but it turns out to be a win.  During `unselect`, we often only need to change a fraction of a Vector's entries, and it is faster to do that using selective `updated` calls  than to rebuild the whole Vector using `map` or `flatMap`. 

> > As the whole time thing is becoming more and more complicated, I think it would be a good a idea to define an "official time policy" for the Zipper to be used as guidelines for future updates.

Are you talking about performance or about the "update time" functionality?  I'm on board with defining both, but I think we're most lacking in the latter case.  Currently, methods like `slice`, `filter` and `unselect` don't yield the intuitive "update time" behavior (nor have they ever).   In this pull request, I fixed `slice` and took a stab at `unselect` because the fixes aligned well with my other modifications.  However, at some point we need to formally specify how update times work across the board.

As for performance, hopefully the new trials are a step in the right direction.  I did run all of the new trials against the original DeepZipper to ensure that we haven't lost anything.   This new code outperforms all earlier implementations.  In some cases the improvement is dramatic.

Even with these changes,  I still think my zipper simplifications were worthwhile.  The basic structure hasn't changed much.  IMO, the ugliest new code is in `slice` and the new `conditionalFlatMapWithIndex`.  The former fixes the "update time" problem I mentioned above.  The latter is mainly for performance, but it only affects `unselect`, and even without that change, the new code's `unselect` was faster than any of the earlier implementations.

> > In the comment for the additionalHoles field you note that it may be used for data associated with actual nodes. As far as I can see, it is always implicitly assumed that the additional data is associated with an empty vector.

The entries in "additionalHoles" are just assertions that the Zipper contains a hole with at least a certain update time.  They don't imply anything about the nodes associated with that hole.   If during `unselect`, we find that an entry in `additionalHoles` doesn't have any nodes associated with it, then we assume the hole is being replaced with an empty vector.  But if it does have nodes associated with it, we retain those nodes and only increase the update time (if needed).

It's true that I only needed `additionalHoles` to account for holes being replaced by empty vectors,  but nothing assumes that fact.   By allowing `additionalHoles` to have "extraneous" entries, I don't have to worry about overlap between `metas` and `additionalHoles` when building the Zipper, and can instead resolve any such overlap during `unselect`.   That yields a net performance gain because most zippers never get unselected.

> > In any case, I think it may be useful in the future to have additional nodes hidden inside the Zipper.

If I understand it correctly, `additionalHoles` probably doesn't satisfy the use you have in mind.  But that can be addressed when it becomes necessary.

josharnold52 committed:
Bloom filter optimizations

josharnold52 committed:
Addressed Daniel B.'s comments.  Cleaned up conditionalFlatMapWithIndex

josharnold52 committed:
Another bloom filter optimization - dont bother checking filter if it is known to be empty

josharnold52 commented:
As discussed, I changed "dead zippers" to no longer use nulls, and moved all of the Zipper's context into a single Option[Context] so now the context is clearly "all or nothing".  I also tried to clarify the "additionalHoles" comments and cleaned-up on some of my self-proclaimed "ugly" code in conditionalFlatMapWithIndex :-)

I also added the bloom filter optimizations I described earlier.   They yield some decent improvements.

@ncreep - Did I address your concerns for this pull?   I figure we can tackle update time specifications and/or performance policies in a separate request.

ncreep commented:
That definitely addresses my concerns, thanks.

When I mentioned "time" before I meant update times. I guess it would be better to tackle it later.

And just for general education: what kind of bugs were there from chained `def`s?

djspiewak commented:
You guys are _awesome!_  Merged.

I'm also curious about the chained defs issue.

josharnold52 commented:
> > I'm also curious about the chained defs issue.

Yeah, that one is worth calling out in the release notes.  Shall I log a separate issue as well?

The problem was that `Zipper.updated` produced a Zipper whose `parent` method delegated to the zipper being updated:

```
new Group(nodes.updated(index, node) with Zipper {
   def parent = self.parent  //"self" refers to the Zipper being updated
   ...
}
```

This meant that new Zipper implicitly retained a reference to the original;   If you made a chain of 100 such updates, the result would indirectly refer to 100 previous instances, and its `parent` method would result in 100 nested `parent` calls.

Only `updated` has this bug;  The other zipper creation methods eagerly evaluate the parent. 

I discovered this when I noticed that Zippers resulting from lots of "updated" calls had very poor unselect performance, due to the nesting of `parent` calls.
